tests/function_modifiers/test_recursive.py

```diff
@@ -539,3 +543,261 @@ def test_recursive_validate_config_inputs_happy(config, inputs):
+    )
+    merge_node = decorator._create_merge_node(upstream_node="upstream_df", node_name="merge_node")
+
+    output_df = merge_node.callable(
+        upstream_df=dummy_df(),
+        dummy_fn_with_columns=dummy_fn_with_columns(col_1=pd.Series([1, 2, 3, 4])),
+    )
+    assert merge_node.name == "merge_node"
+    assert merge_node.type == pd.DataFrame
+
+    pd.testing.assert_series_equal(output_df["col_1"], pd.Series([1, 2, 3, 4]), check_names=False)
+    pd.testing.assert_series_equal(
+        output_df["col_2"], pd.Series([11, 12, 13, 14]), check_names=False
+    )
+    pd.testing.assert_series_equal(
+        output_df["dummy_fn_with_columns"], pd.Series([101, 102, 103, 104]), check_names=False
+    )
+
+
+def test_override_original_column_in_df():
+    def dummy_df() -> pd.DataFrame:
+        return pd.DataFrame({"col_1": [1, 2, 3, 4], "col_2": [11, 12, 13, 14]})
+
+    def target_fn(upstream_df: pd.DataFrame) -> pd.DataFrame:
+        return upstream_df
+
+    def col_1() -> pd.Series:
+        return pd.Series([0, 3, 5, 7])
+
+    decorator = with_columns(col_1, pass_dataframe_as="upstream_df", select="col_1")
+    merge_node = decorator._create_merge_node(upstream_node="upstream_df", node_name="merge_node")
+
+    output_df = merge_node.callable(upstream_df=dummy_df(), col_1=col_1())
+    assert merge_node.name == "merge_node"
+    assert merge_node.type == pd.DataFrame
+
+    pd.testing.assert_series_equal(output_df["col_1"], pd.Series([0, 3, 5, 7]), check_names=False)
+    pd.testing.assert_series_equal(
+        output_df["col_2"], pd.Series([11, 12, 13, 14]), check_names=False
+    )
+
+
+def test_assign_custom_namespace_with_columns():
+    def target_fn(upstream_df: pd.DataFrame) -> pd.DataFrame:
+        return upstream_df
+
+    dummy_node = hamilton_node.Node.from_fn(target_fn)
+    decorator = with_columns(
+        dummy_fn_with_columns,
+        columns_to_pass=["col_1", "col_2"],
+        select="dummy_fn_with_columns",
+        namespace="dummy_namespace",
+    )
+    nodes_ = decorator.transform_dag([dummy_node], {}, target_fn)
+
+    assert nodes_[0].name == "target_fn"
+    assert nodes_[1].name == "dummy_namespace.col_1"
+    assert nodes_[2].name == "dummy_namespace.col_2"
+    assert nodes_[3].name == "dummy_namespace.dummy_fn_with_columns"
+    assert nodes_[4].name == "dummy_namespace.__append"
+
+
+def test_end_to_end_with_columns_automatic_extract():
+    dr = driver.Builder().with_modules(tests.resources.with_columns).build()
+    result = dr.execute(final_vars=["final_df"])["final_df"]
+
+    expected_df = pd.DataFrame(
+        {
+            "col_1": [1, 2, 3, 4],
+            "col_2": [11, 12, 13, 14],
+            "col_3": [1, 1, 1, 1],
+            "subtract_1_from_2": [10, 10, 10, 10],
+            "multiply_3_by_5": [5, 5, 5, 5],
+        }
+    )
+    pd.testing.assert_frame_equal(result, expected_df)
+
+
+def test_end_to_end_with_columns_pass_dataframe():
```

Comment:
Other tests you may want to consider
1. Inputs from outside the subdag
2. Inputs from external stuff
3. Config